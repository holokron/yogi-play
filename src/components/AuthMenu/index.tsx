import * as React from 'react'
import { User } from 'firebase'
import { Nav, NavItem } from 'reactstrap'
import { Link } from 'react-router-dom'
import AuthContainer from '../../containers/AuthContainer'
import { Auth } from '../../lib/auth'

export interface Props {
    onLinkClick?(): void 
}

export default function AuthMenu({onLinkClick = () => {}}: Props): React.ReactElement<Props> {
    return (
        <AuthContainer>
            {(user: User | null, auth: Auth): React.ReactElement<any> => {
                if (user) {
                    return (
                        <Nav navbar>
                            <NavItem>
                                <a href="#" className="nav-link text-uppercase">
                                    {user.email}
                                </a>
                            </NavItem>
                            <NavItem>
                                <a href="#" onClick={auth.logout} className="nav-link text-uppercase">
                                    Wyloguj
                                </a>
                            </NavItem>
                        </Nav>
                    )
                }
            
                return (
                    <Nav navbar>
                        <NavItem>
                            <Link to="/logowanie" className="nav-link text-uppercase" onClick={onLinkClick}>
                                Logowanie
                            </Link>
                        </NavItem>
                        <NavItem>
                            <Link to="/rejestracja" className="nav-link text-uppercase" onClick={onLinkClick}>
                                Rejestracja
                            </Link>
                        </NavItem>
                    </Nav>
                )
            }}
        </AuthContainer>
    )
}