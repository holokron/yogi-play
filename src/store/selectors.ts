import { createSelector } from 'reselect'
import AppState from './state'
import Sound from '../types/Sound'
import SoundsCollection from '../types/SoundsCollection'
import TagsCollection from '../types/TagsCollection'
import Tag from '../types/Tag'

export const getSoundsCollection = createSelector<AppState, SoundsCollection, SoundsCollection>(
    (state: AppState): SoundsCollection => state.sounds,
    (sounds: SoundsCollection): SoundsCollection => sounds,
)

export const getSounds = createSelector<AppState, SoundsCollection, Sound[]>(
    getSoundsCollection,
    (soundsCollection: SoundsCollection): Sound[] => {
        const sounds = Object.values(soundsCollection)

        sounds.sort((soundA: Sound, soundB: Sound): number => {
            const isNewA: boolean = soundA.isNew || false
            const isNewB: boolean = soundB.isNew || false
            const orderA: number = soundA.order || 0
            const orderB: number = soundB.order || 0

            if (isNewA !== isNewB) {
                return isNewA ? -1 : 1
            }

            if (orderA !== orderB) {
                return orderA < orderB ? -1 : 1
            }

            if (soundA.name === soundB.name) {
                return 0
            }

            return soundA.name < soundB.name ? -1 : 1

        })

        return sounds
    }
)

export function createGetSoundsForTag(tag: Tag) {
    return createSelector<AppState, Sound[], Sound[]>(
        getSounds,
        (sounds: Sound[]): Sound[] => sounds
            .filter((sound: Sound) => {
                if ('all' === tag.slug) {
                    return true
                }

                if ('recent' === tag.slug) {
                    return sound.isNew || false
                }

                if ('misc' === tag.slug) {
                    return !sound.tags || !Object.values(sound.tags).length
                }
            
                return sound.tags && sound.tags[tag.id]
            })
    )
}

export function createGetSound(soundId: string) {
    return createSelector<AppState, SoundsCollection, Sound | null>(
        getSoundsCollection,
        (sounds: SoundsCollection): Sound | null => sounds[soundId] || null,
    )
}

export function getSound(state: AppState, soundId: string): Sound | null {
    return state.sounds[soundId] || null
}

export const getTagsCollection = createSelector<AppState, TagsCollection, TagsCollection>(
    (state: AppState): TagsCollection => state.tags,
    (tags: TagsCollection): TagsCollection => tags,
)

export const getTagsByOrder = createSelector<AppState, TagsCollection, Tag[]>(
    getTagsCollection,
    (tagsCollection: TagsCollection): Tag[] => {
        const tags = Object.values(tagsCollection)
        
        tags.sort((tagA: Tag, tagB: Tag) => {
            const orderA = tagA.order || 0
            const orderB = tagB.order || 0

            if (orderA !== orderB) {
                return orderA < orderB ? -1 : 1
            }

            if (tagA.name === tagB.name) {
                return 0
            }

            return tagA.name < tagB.name ? -1 : 1
        })

        return tags
    }
)