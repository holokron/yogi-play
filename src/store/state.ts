import { type SoundsCollection, type TagsCollection, type User } from "@/types";

export default interface AppState {
  sounds: SoundsCollection;
  soundsFilter: string | null;
  tags: TagsCollection;
  chosenTagSlug: string;
  user: User;
  command: {
    open: boolean;
  };
}

export const initialState: AppState = {
  sounds: {},
  soundsFilter: null,
  tags: {},
  chosenTagSlug: "recent",
  user: {
    id: "",
    displayName: "",
    email: "",
  },
  command: {
    open: false,
  },
};
