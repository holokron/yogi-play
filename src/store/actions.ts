import { Action } from 'redux'
import { ThunkAction, ThunkDispatch } from 'redux-thunk'
import * as firebase from 'firebase'
import AppState from './state'
import audioProvider from '../lib/audio-provider'
import database from '../lib/database'
import { getSound } from './selectors'
import Sound from '../types/Sound'
import SoundsCollection from '../types/SoundsCollection'
import TagsCollection from '../types/TagsCollection'
import User from '../types/User'

export enum ACTIONS {
    INITIALIZE = '@app/INITIALIZE',
    PLAY_SOUND = '@app/PLAY_SOUND',
    STOP_SOUND = '@app/STOP_SOUND',
    LOAD_SOUND = '@app/LOAD_SOUND',
    LOAD_SOUNDS = '@app/LOAD_SOUNDS',
    LOAD_TAGS = '@app/LOAD_TAGS',
    CHOOSE_TAG = '@app/CHOOSE_TAG',
    LOAD_USER = '@app/LOAD_USER',
}

const audios: Map<string, HTMLAudioElement> = new Map<string, HTMLAudioElement>()

export interface InitializeAction extends Action<ACTIONS> {}

export interface SoundAction extends Action<ACTIONS> {
    payload: {
        soundId: string
    }
}

export interface LoadSoundsAction extends Action<ACTIONS> {
    payload: {
        sounds: SoundsCollection
    },
}

export interface LoadTagsAction extends Action<ACTIONS> {
    payload: {
        tags: TagsCollection
    },
}

export interface ChooseTagAction extends Action<ACTIONS> {
    payload: {
        tagSlug: string
    },
}

export interface LoadUserAction extends Action<ACTIONS> {
    payload: {
        user: User
    },
}

export type AppAction = SoundAction & LoadSoundsAction & LoadTagsAction & ChooseTagAction & LoadUserAction

export type AppDispatch = ThunkDispatch<AppState, any, AppAction>

export function createInitializeAction(): InitializeAction {
    return {
        type: ACTIONS.INITIALIZE,
    }
}

export function createPlaySoundAction(soundId: string): SoundAction {
    return {
        type: ACTIONS.PLAY_SOUND,
        payload: {
            soundId,
        },
    }
}

export function createStopSoundAction(soundId: string): SoundAction {
    return {
        type: ACTIONS.STOP_SOUND,
        payload: {
            soundId,
        },
    }
}

export function createLoadSoundAction(soundId: string) {
    return {
        type: ACTIONS.LOAD_SOUND,
        payload: {
            soundId,
        },
    }
}

export function createLoadSoundsAction(sounds: SoundsCollection): LoadSoundsAction {
    return {
        type: ACTIONS.LOAD_SOUNDS,
        payload: {
            sounds,
        },
    }
}

export function createLoadTagsAction(tags: TagsCollection): LoadTagsAction {
    return {
        type: ACTIONS.LOAD_TAGS,
        payload: {
            tags,
        },
    }
}

export function createChooseTagAction(tagSlug: string): ChooseTagAction {
    return {
        type: ACTIONS.CHOOSE_TAG,
        payload: {
            tagSlug,
        },
    }
}

export function createLoadUserAction(user: User): LoadUserAction {
    return {
        type: ACTIONS.LOAD_USER,
        payload: {
            user,
        },
    }
}

export function playSound(soundId: string): ThunkAction<void, AppState, any, SoundAction> {
    return async (dispatch, getState) => {
        const sound: Sound | null = getSound(getState(), soundId)

        if (!sound) {
            return
        }

        let audio: HTMLAudioElement | undefined = audios.get(soundId)

        if (!audio) {                
            const soundSrc: string = await audioProvider.getAudio(sound.path)
            audio = new Audio(soundSrc)
            audio.preload = 'auto'
            audio.addEventListener('loadstart', (): void => {
                dispatch(createLoadSoundAction(soundId))
            })
            audio.addEventListener('ended', (): void => {
                dispatch(createStopSoundAction(soundId))
            })
            audio.addEventListener('pause', (): void => {
                dispatch(createStopSoundAction(soundId))
            })
            audio.addEventListener('playing', (): void => {
                dispatch(createPlaySoundAction(soundId))
            })

            audios.set(soundId, audio)
        }

        audio.play()
    }
}

export function stopSound(soundId: string):  ThunkAction<void, AppState, any, SoundAction> {
    return (dispatch) => {
        let audio: HTMLAudioElement | undefined = audios.get(soundId)

        if (!audio) {
            return
        }

        audio.pause()
        audio.currentTime = 0

        dispatch(createStopSoundAction(soundId))
    }
}

export function loadSounds(): ThunkAction<void, AppState, any, LoadSoundsAction> {
    return (dispatch) => {        
        database.getSoundsRef()
            .on(
                'value',
                (snapshot: firebase.database.DataSnapshot) => {
                    dispatch(createLoadSoundsAction(snapshot.val()))
                }
            )
    }
}

export function loadTags(): ThunkAction<void, AppState, any, LoadTagsAction> {
    return (dispatch) => {
        database.getTagsRef()
            .on(
                'value',
                (snapshot: firebase.database.DataSnapshot) => {
                    dispatch(createLoadTagsAction(snapshot.val()))
                }
            )
    }
}

export function chooseTag(tagSlug: string): ThunkAction<void, AppState, any, ChooseTagAction> {
    return (dispatch) => {
        dispatch(createChooseTagAction(tagSlug))
    }
}

export function readText(text: string): ThunkAction<void, AppState, any, any> {
    return () => {
        if (!speechSynthesis) {
            return
        }

        const speech: SpeechSynthesisUtterance = new SpeechSynthesisUtterance(text)
        speech.rate = 0.7
        speechSynthesis.speak(speech)
    }
}

export function authenticate(): ThunkAction<void, AppState, any, LoadUserAction> {
    return (dispatch) => {
        let userId: string | null = localStorage.getItem('userId')

        if (!userId) {

            console.log('no userId in localstorage, signing in anonymously')

            firebase.auth().signInAnonymously()
                .then((value: firebase.auth.UserCredential) => {
                    if (!value.user) {
                        console.log('no user credentials')

                        return
                    }

                    console.log(`setting userId: ${value.user.uid} in localstorage`, value)

                    localStorage.setItem('userId', value.user.uid)
                })
        }

        firebase.auth().onAuthStateChanged(async (user: firebase.User | null): Promise<void> => {
            if (!user) {
                console.log('no authenticated')

                return
            }

            console.log(`getting user: ${user.uid} from database`)

            firebase.database().ref(`/users/${user.uid}`)
                .on('value', (data: firebase.database.DataSnapshot) => {

                    firebase.auth().fetchSignInMethodsForEmail(data.val().email).then(methods => console.log(methods))

                    console.log(`got user: ${user.uid} from database`)

                    console.log(data.val())

                    dispatch(createLoadUserAction(data.val()))
                })

        })
    }
}