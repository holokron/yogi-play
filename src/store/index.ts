import { applyMiddleware, compose, createStore, Store } from "redux";
import { thunk, ThunkDispatch } from "redux-thunk";
import { composeWithDevTools } from "@redux-devtools/extension";
import appReducer from "./reducer";
import AppState, { initialState } from "./state";
import { AppAction } from "./actions";
import { config } from "../lib/config";

const middleware = applyMiddleware(thunk);

const enhancer =
  config.nodeEnv === "development"
    ? composeWithDevTools(middleware)
    : compose(middleware);

export type AppDispatch = ReturnType<typeof configureStore>["dispatch"];
export type AppThunkDispatch = ThunkDispatch<AppState, undefined, AppAction>;

export default function configureStore(
  state: AppState = initialState
): Store<AppState, AppAction> {
  const store = createStore(appReducer, state, enhancer);

  return store;
}
