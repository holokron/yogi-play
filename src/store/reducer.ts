import AppState, { initialState } from "./state";
import { ACTIONS, AppAction } from "./actions";

export default function appReducer(
  state: AppState = initialState,
  action: AppAction
): AppState {
  switch (action.type) {
    case ACTIONS.PLAY_SOUND:
      return {
        ...state,
        soundStates: {
          ...state.soundStates,
          [action.payload.soundId]: {
            ...state.soundStates[action.payload.soundId],
            soundId: action.payload.soundId,
            isPlaying: true,
            isLoading: false
          }
        }
      };
    case ACTIONS.STOP_SOUND:
      return {
        ...state,
        soundStates: {
          ...state.soundStates,
          [action.payload.soundId]: {
            ...state.soundStates[action.payload.soundId],
            soundId: action.payload.soundId,
            isPlaying: false,
            isLoading: false
          }
        }
      };
    case ACTIONS.LOAD_SOUND:
      return {
        ...state,
        sounds: {
          ...state.sounds,
          [action.payload.soundId]: {
            ...state.sounds[action.payload.soundId],
            isLoading: true
          }
        }
      };
    case ACTIONS.LOAD_SOUNDS:
      return {
        ...state,
        sounds: {
          ...state.sounds,
          ...action.payload.sounds
        }
      };
    case ACTIONS.LOAD_TAGS:
      return {
        ...state,
        tags: {
          ...state.tags,
          ...action.payload.tags
        }
      };
    case ACTIONS.CHOOSE_TAG:
      return {
        ...state,
        chosenTagSlug: action.payload.tagSlug
      };
    case ACTIONS.LOAD_USER:
      return {
        ...state,
        user: action.payload.user
      };
    case ACTIONS.ADD_USER_SOUND:
      return {
        ...state,
        user: {
          ...state.user,
          sounds: {
            ...state.user.sounds,
            [action.payload.soundId]: true
          }
        }
      };
    default:
      return state;
  }
}
