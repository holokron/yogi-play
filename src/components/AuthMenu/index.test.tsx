import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import AuthMenu from ".";
import configureStore from "../../store";
import { initialState } from "../../store/state";

const store = configureStore({
  ...initialState,
  user: {
    id: "foo",
    displayName: "Foo Bar",
    email: "foo@bar.com",
  },
});

describe("@components/AuthMenu", () => {
  it("renders correctly", () => {
    const tree = render(
      <Provider store={store}>
        <MemoryRouter>
          <AuthMenu />
        </MemoryRouter>
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});
