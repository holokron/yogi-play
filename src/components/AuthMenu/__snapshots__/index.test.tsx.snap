// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/AuthMenu renders correctly 1`] = `
<ul
  className="navbar-nav"
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link text-uppercase"
      href="/logowanie"
      onClick={[Function]}
    >
      Logowanie
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link text-uppercase"
      href="/rejestracja"
      onClick={[Function]}
    >
      Rejestracja
    </a>
  </li>
</ul>
`;
